### PR TITLE
PMSO-678: Create Delivery trips per order

### PR DESCRIPTION
**Pull request by LinhNguyen310**
### Description
This PR introduces the ProductOrderDeliveryTripfeature, which allows the creation, validation, and management of delivery trips associated with product orders. The ProductOrderDeliveryTripSerialize ensures that delivery trips are properly validated and linked to existing product orders and products.
Please include a summary of the changes and the related issue(s) this PR addresses.

### Related Jira(s)

- [PMSO-678](https://trietpham.atlassian.net/jira/software/projects/PMSO/boards/6?selectedIssue=PMSO-678)

### Checklist

- [ ] I have read and followed the [PR naming convention](https://pulsar.apache.org/contribute/develop-semantic-title/)
- [ ] My code follows the code style of this project.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### Tests
Test Adding a Delivery Trip:

1. Verified that a delivery trip can be successfully created with valid product_order and product IDs.
2. Ensured that the response includes all expected fields, such as car_plate, trip_order, and quantity.

Test Listing Delivery Trips:

1. Confirmed that all delivery trips associated with a specific product_order are returned correctly.

Test Validation:

1. Ensured that an error is raised when attempting to add a delivery trip with an invalid product_order or product.

Test Edge Cases:

1. Verified behavior when no product_order is provided.
2. Tested scenarios with duplicate trip_order values for the same product_order.

[PMSO-678]: https://trietpham.atlassian.net/browse/PMSO-678?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: pmsoService/products/models.py [hunk 1]
```diff
@@ -193,6 +188,10 @@ class ProductOrder(models.Model):
         null=True,
         related_name="storage_keeper_orders",
     )
+
+    product_order = models.ForeignKey(
```
Comment (phamhuutriet) [inline pmsoService/products/models.py:192]: field này để làm gì v e?

File: pmsoService/products/models.py [hunk 2]
```diff
@@ -372,3 +366,30 @@ class Meta:
...
+    )
+    car_plate = models.CharField(max_length=20)
+    date = models.BigIntegerField()
+    status = models.CharField(
+        max_length=20,
+        choices=DELIVERY_TRIP_STATUS_CHOICES,
+        default='pending',
+    )
+    returned_quantity = models.FloatField(default=0, blank=True, null=True)
+    delivery_products = models.JSONField(default=list)  # Store delivery products as a list of dictionaries
```
Comment (phamhuutriet) [inline pmsoService/products/models.py:386]: không nên làm như này vì nếu thông tin về product thay đổi sẽ không sync được. Chưa kể mình sẽ có view aggregate chuyến vận chuyển giữa những product order với nhau. Trong trường hợp này, phải tạo 1 relationship table ProductOrderDeliveryProduct để lưu list of products

**Review by phamhuutriet — CHANGES_REQUESTED** [review]

File: pmsoService/products/models.py (hunk 1 above)
Comment (LinhNguyen310) [reply pmsoService/products/models.py:192]: ah cái này nên là product_order_delivery_trips ạ, thanks anh em chắc rename mà không để ý

File: pmsoService/products/models.py (hunk 2 above)
Comment (LinhNguyen310) [reply pmsoService/products/models.py:386]: ah dạ để em update ạ

File: pmsoService/products/tests/test_delivery.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    def setUp(self):
+        # Create a sample customer
+        self.customer = Customer.objects.create(
+            id=uuid.uuid4(),
+            name="Test Customer",
+            email="testcustomer@example.com",
+        )
+
+        # Create a sample product order
+        self.product_order = ProductOrder.objects.create(
```
Comment (phamhuutriet) [inline pmsoService/products/tests/test_delivery.py:17]: em xem những test ở product order cách set up data: https://github.com/associatedG/pmso-service/blob/staging/pmsoService/products/tests/product_order/test_create.py. Follow theo pattern set up context trong những test này. Ưu tiên dùng defined helper function với Factory class để generate entity

File: pmsoService/products/tests/test_delivery.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+                            "quantity": 111,
+                            "name": "abc"
+                        }
+                    ]
+                }
+            ]
+        }
+
+        # Send PATCH request
+        response = self.client.patch(
```
Comment (phamhuutriet) [inline pmsoService/products/tests/test_delivery.py:72]: extract cái này ra thành 1 function riêng, refer theo test đính kèm ở trên để xem pattern

File: pmsoService/products/tests/test_delivery.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        delivery_trip = delivery_trips.first()
+        self.assertEqual(delivery_trip.car_plate, "1233")
+        self.assertEqual(delivery_trip.status, "Processing")
+        self.assertEqual(delivery_trip.returned_quantity, 0)
+
+        # Assert the delivery products were updated correctly
+        delivery_products = delivery_trip.delivery_products.all()
+        self.assertEqual(delivery_products.count(), 2)
+
+        product1 = delivery_products.get(product=self.product1)
```
Comment (phamhuutriet) [inline pmsoService/products/tests/test_delivery.py:95]: rename product1 lại thành `deliver_product1` để biết đây là product bên trong delivery trip

File: pmsoService/products/tests/test_delivery.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        self.assertEqual(product1.quantity, 10)
+        self.assertEqual(product1.name, "2")
+
+        product2 = delivery_products.get(product=self.product2)
+        self.assertEqual(product2.quantity, 111)
+        self.assertEqual(product2.name, "abc")
```
Comment (phamhuutriet) [inline pmsoService/products/tests/test_delivery.py:101]: thiếu test cho create delivery trip

File: pmsoService/products/views.py [hunk 7]
```diff
@@ -168,7 +168,7 @@ def get_filterset(self, queryset=None):
 
 
 class ProductOrderRetrieveUpdateDestroyView(generics.RetrieveUpdateDestroyAPIView):
-    queryset = ProductOrder.objects.all()
+    queryset = ProductOrder.objects.prefetch_related('delivery_trips')
```
Comment (phamhuutriet) [inline pmsoService/products/views.py:171]: cái này để chi vậy e?

File: pmsoService/products/views.py [hunk 8]
```diff
@@ -177,6 +177,20 @@ def get_serializer_class(self):
             return GetProductOrderSerializer
         return ProductOrderSerializer
 
+    def update(self, request, *args, **kwargs):
```
Comment (phamhuutriet) [inline pmsoService/products/views.py:180]: tại sao lại implement function này? Bên trong ProductOrderSerializer đã có update function rồi mà?

File: pmsoService/utils/choices_utils.py [hunk 9]
```diff
@@ -137,3 +137,31 @@ def get_storage_keeper_role_mappings():
...
+def get_all_storage_choices():
+    """
+    Get all Storage of Material
+    """
+    choices_config_file = "config/choices_config.json"
+    storage = read_file(choices_config_file)
+    return storage["STORAGE"]["STORAGE_CHOICES"]
+
+
+def get_storage_keeper_role_mappings():
```
Comment (phamhuutriet) [inline pmsoService/utils/choices_utils.py:151]: đã có function này trong role_utils rồi

File: pmsoService/products/serializers.py [hunk 10]
```diff
@@ -449,14 +531,25 @@ def update(self, instance, validated_data):
...
+        delivery_trips_data = validated_data.pop("delivery_trips", None)
+        instance = super().update(instance, validated_data)  # Update the main instance
+
+        if delivery_trips_data is not None:
+            process_delivery_trips(instance, delivery_trips_data)
+
         # Process product data
         products_data = validated_data.pop("products", None)
-        return update_product_order(instance, validated_data, products_data)
+        if products_data is not None:
```
Comment (phamhuutriet) [inline pmsoService/products/serializers.py:549]: chỗ này không đúng, với logic này thì nếu update những field khác mà không update product data thì sẽ không update được

File: pmsoService/products/serializers.py [hunk 11]
```diff
@@ -14,13 +14,93 @@
...
 from rest_framework.validators import UniqueValidator
+from .services.product_order_delivery_trip_services import *
 
 
 def get_field_permissions():
     permissions_config = read_file("config/permissions_config.json")
     return permissions_config.get("field_permissions", {})
 
 
+class ProductSerializer(ProductValidationMixin, serializers.ModelSerializer):
```
Comment (phamhuutriet) [inline pmsoService/products/serializers.py:25]: hiện tại anh đang thấy có 2 ProductSerializer trong file này, remove cái còn lại

File: pmsoService/products/views.py [hunk 12]
```diff
@@ -1036,3 +1050,56 @@ def get_customer_summary(request):
         )
 
     return JsonResponse(data, safe=False)
+
+class ProductOrderDeliveryTripViewSet(viewsets.ModelViewSet):
```
Comment (phamhuutriet) [inline pmsoService/products/views.py:1054]: view này để làm gì vậy em?

File: pmsoService/products/views.py [hunk 13]
```diff
@@ -1036,3 +1050,56 @@ def get_customer_summary(request):
...
+    def get_queryset(self):
+        product_order_id = self.kwargs.get("product_order_id")
+        return ProductOrderDeliveryTrip.objects.filter(product_order_id=product_order_id)
+
+    def perform_update(self, serializer):
+        self.request.data.get('delivery_products', [])
+        serializer.save()
+
+
+class ProductOrderDeliveryTripListCreateView(ListCreateAPIView):
```
Comment (phamhuutriet) [inline pmsoService/products/views.py:1092]: hiện tại anh đang thấy có sự không nhất quán trong cách em tạo delivery trip. Nếu em tạo 2 view này và có API riêng cho delivery trip thì tại sao lại tạo delivery trip thông qua patch Product Order? Nên chọn 1 trong 2 thôi: 1 là CRUD delivery trip thông qua API của riêng nó, hoặc là chỉ làm thông qua patch ProductOrder thôi

**Review by phamhuutriet — CHANGES_REQUESTED** [review]
build fail do migration đang bị lỗi, em có xoá migration file nào so với những commit trước đó không?

File: pmsoService/products/models.py [hunk 14]
```diff
@@ -379,3 +389,52 @@ class Meta:
...
+        blank=True,
+    )
+    car_plate = models.CharField(max_length=20)
+    date = models.DateTimeField(default=timezone.now)
+    status = models.CharField(
+        max_length=20,
+        choices=DELIVERY_TRIP_STATUS_CHOICES,
+        default="pending",
+    )
+    returned_quantity = models.FloatField(default=0, blank=True, null=True)
```
Comment (phamhuutriet) [inline pmsoService/products/models.py:409]: field này đã có trong ProductOrderDeliveryProduct để track returned quantity của từng sp rồi vậy field này ở đây để làm gì v e?

File: pmsoService/products/models.py [hunk 15]
```diff
@@ -379,3 +389,52 @@ class Meta:
...
+        "ProductOrderDeliveryTrip",
+        on_delete=models.CASCADE,
+        related_name="delivery_products",
+    )
+    product = models.ForeignKey(
+        "Product", on_delete=models.CASCADE, related_name="delivery_products"
+    )
+    quantity = models.FloatField(default=0)
+    returned_quantity = models.FloatField(default=0)
+    name = models.CharField(max_length=255, blank=True, null=True)
```
Comment (phamhuutriet) [inline pmsoService/products/models.py:433]: name này là tên của gì v em?

File: pmsoService/products/serializers.py [hunk 16]
```diff
@@ -14,13 +14,94 @@
...
+    delivery_products = ProductOrderDeliveryProductSerializer(many=True)
+
+    class Meta:
+        model = ProductOrderDeliveryTrip
+        fields = [
+            "id",
+            "car_plate",
+            "date",
+            "status",
+            "returned_quantity",
```
Comment (phamhuutriet) [inline pmsoService/products/serializers.py:99]: giống như comment trên, field này để làm gì vậy em?

File: pmsoService/products/services/product_order_delivery_trip_services.py [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+                process_trip_update(delivery_trip, trip_data)
+                to_update.append(delivery_trip)
+                to_delete_ids.remove(trip_id)
+            else:
+                to_create_trips_data.append((trip_data, delivery_products_data))
+
+    return to_update, to_delete_ids, to_create_trips_data
+
+
+def process_product_updates(existing_products, products_data):
```
Comment (phamhuutriet) [inline pmsoService/products/services/product_order_delivery_trip_services.py:49]: method này dùng ở đâu v em?

File: pmsoService/products/tests/product_order_delivery_trip/test_create.py [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+
+    def test_create_delivery_trip(self):
+        product_order = self.create_product_order()
+        product = self.create_product()
+
+        delivery_product = self.delivery_product_body.copy()
+        delivery_product["product_id"] = str(product.id)
+
+        request_body = self.delivery_trip_body.copy()
+        request_body["id"] = str(uuid.uuid4())
```
Comment (phamhuutriet) [inline pmsoService/products/tests/product_order_delivery_trip/test_create.py:27]: id là read-only field nên không cần bỏ vô request_body

File: pmsoService/products/tests/product_order_delivery_trip/test_create.py [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+            self.assertEqual(response_data["delivery_products"][i]["returned_quantity"], delivery_product["returned_quantity"])
+
+    def test_create_delivery_trip(self):
+        product_order = self.create_product_order()
+        product = self.create_product()
+
+        delivery_product = self.delivery_product_body.copy()
+        delivery_product["product_id"] = str(product.id)
+
+        request_body = self.delivery_trip_body.copy()
```
Comment (phamhuutriet) [inline pmsoService/products/tests/product_order_delivery_trip/test_create.py:26]: không cần copy vì setUp sẽ được gọi trước mỗi test nên data sẽ reset lại

File: pmsoService/products/tests/product_order_delivery_trip/test_create.py [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+                delivery_product["product_id"] = str(product.id)
+                delivery_products.append(delivery_product)
+
+            delivery_trip = self.delivery_trip_body.copy()
+            delivery_trip["id"] = str(uuid.uuid4()) 
+            delivery_trip["date"] = f"2025-03-25T14:00:0{i}Z"
+            delivery_trip["delivery_products"] = delivery_products
+            delivery_trips.append(delivery_trip)
+
+        response = self.patch_product_order(
```
Comment (phamhuutriet) [inline pmsoService/products/tests/product_order_delivery_trip/test_create.py:77]: nên extract cái này thành `create_delivery_trips` method luôn

File: pmsoService/products/tests/product_order_delivery_trip/test_detail.py [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+        request_body["date"] = "2025-03-25T14:00:00Z"
+        request_body["delivery_products"] = [delivery_product]
+
+        response = self.patch_product_order(
+            product_order.id, {"delivery_trips": [request_body]}
+        )
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+        delivery_trip_id = response.data["delivery_trips"][0]["id"]
+        response = self.client.get(
```
Comment (phamhuutriet) [inline pmsoService/products/tests/product_order_delivery_trip/test_detail.py:26]: tương tự như trên, extract nguyên khúc này thành 1 function `get_delivery_trips`

File: pmsoService/products/tests/product_order_delivery_trip/test_create.py [hunk 22]
```diff
@@ -0,0 +1,85 @@
+import uuid
+from rest_framework import status
+from .test_base import TestDeliveryTripAPIBase
+
+
+class TestDeliveryTripCreate(TestDeliveryTripAPIBase):
```
Comment (phamhuutriet) [inline pmsoService/products/tests/product_order_delivery_trip/test_create.py:6]: nên thêm 1 số test nếu như tạo delivery trip with invalid products, date time

File: pmsoService/products/tests/product_order_delivery_trip/test_list.py [hunk 23]
```diff
@@ -0,0 +1,62 @@
...
+        products = [self.create_product() for _ in range(2)]
+
+        response = self.client.get(
+            reverse("product_order_detail", kwargs={"id": product_order.id})
+        )
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        existing_delivery_trips = response.data.get("delivery_trips", [])
+
+        delivery_trips = []
+        for i in range(3):
```
Comment (phamhuutriet) [inline pmsoService/products/tests/product_order_delivery_trip/test_list.py:19]: khúc set up này nên dùng factory, xem link anh đã dẫn ở những comment trước. Tạo 1 class DeliveryTripFactory rồi tạo 1 helper method để tạo DeliveryTrip thẳng trong database luôn. Nó sẽ trông ntn:

```
for i in range(num_of_trips):
     self.create_delivery_trip(name=name, date=date)
```

sau đó gọi get method để lấy full list và assert

File: pmsoService/products/tests/product_order_delivery_trip/test_update.py [hunk 24]
```diff
@@ -0,0 +1,143 @@
...
+
+class TestDeliveryTripUpdate(TestDeliveryTripAPIBase):
+    def test_update_delivery_trip_status(self):
+        product_order = self.create_product_order()
+        product = self.create_product()
+
+        delivery_product = self.delivery_product_body.copy()
+        delivery_product["product_id"] = str(product.id)
+
+        request_body = self.delivery_trip_body.copy()
```
Comment (phamhuutriet) [inline pmsoService/products/tests/product_order_delivery_trip/test_update.py:14]: nếu test nào cũng tạo delivery trip y như nhau thì nên tạo sẵn ở setUp method và chỉ patch những cái field muốn test trong từng test thôi

**Review by phamhuutriet — CHANGES_REQUESTED** [review]
Good progress !!

**Review by phamhuutriet — DISMISSED** [review]
PR ổn r nha, resolve merge conflict là được

**Review by phamhuutriet — APPROVED** [review]
